### ISSUE TITLE
snapshot_name doesn't allow colon's

### ISSUE THREAD

**Issue by samcv** [digint/btrbk#213]
I cannot set snapshot_name to `var:opt` it says `ERROR: Ambiguous file for option "snapshot_name" in "/etc/btrbk/btrbk.conf" line 130: var:opt`. If this is a purposeful limitation then it should be in the man page.

```
# Backup to external disk mounted on /mnt/btr_backup
volume /mnt/btr_pool
  # no action if external disk is not attached
  snapshot_create  ondemand
  target send-receive /mnt/btr_backup/_btrbk
  subvolume var/opt
    snapshot_name var:opt
  subvolume srv
  subvolume var/tmp
    snapshot_name var:tmp
```

**Comment by digint** [comment]
Yes, this is a "purposeful" limitation. It's in [btrbk.conf(5)](https://digint.ch/btrbk/doc/btrbk.conf.5.html):
"Note that btrfs is very picky on file names (mainly for security reasons), only the characters [0-9] [a-z] [A-Z] and "._+-@" are allowed."

Actually, this applies to *ALL* filenames processed by btrbk:

```
# git grep accept_file
btrbk:  snapshot_dir                => { default => undef,     accept_file => { relative => 1 }, context => [ "root", "volume", "subvolume" ] },
btrbk:  snapshot_name               => { c_default => 1,       accept_file => { name_only => 1 }, context => [ "subvolume" ], deny_glob_context => 1 },   # NOTE: defaults to the subvolume name (hardcoded)
btrbk:  ssh_identity                => { default => undef,     accept_file => { absolute => 1 } },
btrbk:  transaction_log             => { default => undef,     accept => [ "no" ], accept_file => { absolute => 1 }, context => [ "root" ] },
btrbk:  lockfile                    => { default => undef,     accept => [ "no" ], accept_file => { absolute => 1 }, context => [ "root" ] },
btrbk:  gpg_keyring                 => { default => undef,     accept_file => { absolute => 1 } },
btrbk:  openssl_keyfile             => { default => undef,     accept_file => { absolute => 1 } },
btrbk:  kdf_backend                 => { default => undef,     accept_file => { absolute => 1 } },
btrbk:  archive_exclude             => { default => undef,     accept_file => { wildcards => 1 }, allow_multiple => 1, context => [ "root" ] },
```

Maybe it's not prominent enough in "SECTIONS". Do you have any suggestions where I should put it? (I definitively don't want to repeat this on every config option). Maybe a "LIMITATIONS" section?

There are pretty complex shell commands called with filenames in btrbk, and escaping them correctly is complex and often varies in different shells. Most special characters are regarded as evil, as this can lead to severe security issues, especially as btrbk needs root in most cases. So I tend to be very strict about this, as I can't test every OS/shell used by btrbk users.
For example, the colon ":" character is often used to separate filenames (as in PATH=/bin:/usr/bin), it's probably not a good idea to have files named like this in the first place (you could go even further and create files named like: "let's have a $dollar signs in filenames, what could possibly go wrong?").

If you want to enable ":", change the regex in [btrbk line 60/61](/digint/btrbk/blob/v0.26.1/btrbk#L60) to:
```
my $file_match = qr/[0-9a-zA-Z:_@\+\-\.\/]+/;  # note: ubuntu uses '@' in the subvolume layout: <https://help.ubuntu.com/community/btrfs>
my $glob_match = qr/[0-9a-zA-Z:_@\+\-\.\/\*]+/; # file_match plus '*'
```
This should/might work (untested).

**Comment by samcv** [comment]
What I would have expected is a message telling me that I had used invalid characters in my backup name. That would be the most user friendly. The limitations section does sound good, as I did look there after I got that error message and had no clue what was causing it.

**Comment by samcv** [comment]
@digint it seems I can create a volume called `!@#$%^&*()[]{}\` so btrfs dosen't seem to disallow those, or the : sign.

Also btrbk does not allow any non-ascii alphanum, which could be an issue for people who don't use English. I would think there shouldn't be too many security concerns if it kept the certain whitelist for ascii but allowed non-ascii unicode. This could possibly be hidden behind a config option. What are your thoughts on that? I'd be willing to work on that and submit a pull request if that's something you are ok with.

Another thing i was thinking about is a config setting to set a custom character whitelist or blacklist, so btrbk can be usable in more use case scenarios, and a configurable option to allow unicode characters as well. If we wanted we could even restrict it just to alphabetic unicode characters if you were concerned about allowing unicode symbols through. What do you think?

**Comment by digint** [comment]
I think best would be an option (something like `insecure_paths=yes`) which would allow unicode as well as special chars like `!@#$%^&*()[]{}\`. If escaped correctly everywhere, there should be no issue. This way, people who really want this can enable it, and file bug reports accordingly. As soon as this stabilizes, we can make it the default without breaking anything. Defining a "custom character whitelist or blacklist" in a config option is probably not a good idea, because if we really need to blacklist some characters, there are good reasons for it.

@samcv if you start working on a patch for this, please use the **snapdir-mountpoint-v2 branch**, which contains many changes on how paths are processed (this should work, not yet merged as it needs some more testing on several platforms).

Here's a list of things that might go wrong with unicode and special chars:

- ssh commands: these definitively need to be escaped correctly. This can get hard as they need to be 
escaped twice (shell calling ssh, which again spawns a shell on the remote side)
- system commands: btrbk uses a bunch of unix and btrfs commands
  - `readlink -e <file>`
  - `cat /proc/self/mounts` (which has its own encoding I believe)
  - `btrfs subvolume list <file>` (afaik this does not escape anything, but simply prints the file name at the end of a line)
  - `btrfs subvolume show <file>`
  - some more, especially when using raw targets
- missing system locale: someone uses a backup disk with unicode snapshot names on a system supporting only the "C" locale.
- output of `btrbk --format=raw` needs well-defined escaping
- `ssh_filter_btrbk.sh`: this is for paranoid people, and it's probably fine if we don't allow special chars here
- perl: handles unicode "just right", but there also might be some corner cases
- ... probably some more...

*beware, localization is a bitch*...

**Comment by digint** [comment]
Implemented in btrbk-0.32.0, closing issue
